Dockerfile

```diff
@@ -1,71 +1,74 @@
 
-# Install node packages defined in package.json, including webpack
 COPY package.json yarn.lock /app/
 RUN yarn install --frozen-lockfile
 
-# Copy all files to /app (except what is defined in .dockerignore)
 COPY . /app/
 
-# Compile assets and run webpack. We set a dummy secret key.
-RUN RAILS_ENV=production bundle exec rails SECRET_KEY_BASE=pickasecuretoken assets:precompile
+RUN bundle exec rake assets:precompile
 
-# Cleanup to save space in the production image
-RUN rm -rf node_modules log tmp && \
-      rm -rf /usr/local/bundle/cache && \
-      rm -rf .env && \
-      find /usr/local/bundle/gems -name "*.c" -delete && \
-      find /usr/local/bundle/gems -name "*.h" -delete && \
-      find /usr/local/bundle/gems -name "*.o" -delete && \
-      find /usr/local/bundle/gems -name "*.html" -delete
+ENV PORT=8080
+EXPOSE ${PORT}
 
-# Build runtime image
-FROM ruby:3.1.4-alpine as production
+RUN adduser --system --no-create-home nonroot
 
-# The application runs from /app
-WORKDIR /app
+FROM base as test
 
-# Add postgres driver library
-# Add the timezone as it's not configured by default in Alpine
-RUN apk add --update --no-cache libpq tzdata && \
-  cp /usr/share/zoneinfo/Europe/London /etc/localtime && \
-  echo "Europe/London" > /etc/timezone
+RUN bundle config set without ""
+RUN bundle install --jobs=4 --no-binstubs --no-cache
 
-# Copy files generated in the builder image
-COPY --from=builder /app /app
-COPY --from=builder /usr/local/bundle/ /usr/local/bundle/
+# Install gecko driver for Capybara tests
+RUN apk add firefox
+RUN wget https://github.com/mozilla/geckodriver/releases/download/v0.31.0/geckodriver-v0.31.0-linux64.tar.gz \
+    && tar -xvzf geckodriver-v0.31.0-linux64.tar.gz \
+    && rm geckodriver-v0.31.0-linux64.tar.gz \
+    && chmod +x geckodriver \
+    && mv geckodriver /usr/local/bin/
 
-ARG GIT_COMMIT_SHA="UNKNOWN"
-ENV GIT_COMMIT_SHA=${GIT_COMMIT_SHA}
-RUN echo ${GIT_COMMIT_SHA} > ./GIT_COMMIT_SHA
+CMD bundle exec rake parallel:setup && bundle exec rake parallel:spec
 
-ENV PORT=8080
+FROM base as development
 
-EXPOSE ${PORT}
+# We expect the rake assets:precompile command to create these directories, but mkdir -p will create them if they don't already exist
+RUN mkdir -p tmp log
+RUN chown -R nonroot tmp log
+RUN chown nonroot db/schema.rb
+
+RUN bundle config set without "test"
+RUN bundle install --jobs=4 --no-binstubs --no-cache
+
+USER nonroot
+
+CMD bundle exec rails s -e ${RAILS_ENV} -p ${PORT} --binding=0.0.0.0
+
+FROM base as staging
+
+# We expect the rake assets:precompile command to create these directories, but mkdir -p will create them if they don't already exist
+RUN mkdir -p tmp log
+RUN chown -R nonroot tmp log
+RUN chown nonroot db/schema.rb
+
+USER nonroot
+
+CMD bundle exec rails s -e ${RAILS_ENV} -p ${PORT} --binding=0.0.0.0
 
-CMD RAILS_ENV=${RAILS_ENV} bundle exec rake db:migrate && bundle exec rails s -e ${RAILS_ENV} -p ${PORT} --binding=0.0.0.0
+FROM staging as production
```

Comment:
What's the purpose of this? I think it's neater to build and publish a single image for deployment to staging/production.

You've set this up to let you `docker build` with both `--target staging` and `--target production` but they're actually the same?

Reply:
My thinking was it would indicate the Dockerfile can build images for `staging` and `production`, and that they're the same. But I agree it would be neater to just have a `production` stage to serve both, and maybe its easy enough to decipher anyway, so I'll make this change 😀